apply-templates.sh

```diff
@@ -0,0 +1,83 @@
+#!/usr/bin/env bash
+set -Eeuo pipefail
+
+[ -f versions.json ] # run "versions.sh" first
+
+jqt='.jq-template.awk'
+if [ -n "${BASHBREW_SCRIPTS:-}" ]; then
+	jqt="$BASHBREW_SCRIPTS/jq-template.awk"
+elif [ "$BASH_SOURCE" -nt "$jqt" ]; then
+	# https://github.com/docker-library/bashbrew/blob/master/scripts/jq-template.awk
+	wget -qO "$jqt" 'https://github.com/docker-library/bashbrew/raw/9f6a35772ac863a0241f147c820354e4008edf38/scripts/jq-template.awk'
+fi
+
+if [ "$#" -eq 0 ]; then
+	versions="$(jq -r 'keys | map(@sh) | join(" ")' versions.json)"
+	eval "set -- $versions"
+fi
+
+generated_warning() {
+	cat <<-EOH
+		#
+		# NOTE: THIS DOCKERFILE IS GENERATED VIA "apply-templates.sh"
+		#
+		# PLEASE DO NOT EDIT IT DIRECTLY.
+		#
+
+	EOH
+}
+
+for version; do
+	export version
+
+	rm -rf "$version"
+
+	if jq -e '.[env.version] | not' versions.json > /dev/null; then
+		echo "deleting $version ..."
+		continue
+	fi
+
+	variants="$(jq -r '.[env.version].variants | map(@sh) | join(" ")' versions.json)"
+    eval "variants=( $variants )"
+
+    phpVersion="$(jq -r '.[env.version] | .version' versions.json)"
+    eval "phpVersion=( $phpVersion )"
+
+	for dir in "${variants[@]}"; do
+		suite="$(dirname "$dir")" # alpine
+		variant="$(basename "$dir")" # fpm-nginx
+        phpMajorVersion="${version//.}" # 80
+        export suite variant phpMajorVersion phpVersion
+
+		alpineVer="${suite#alpine}" # "3.12", etc
+		if [ "$suite" != "$alpineVer" ]; then
+			from="php:$phpVersion-fpm-alpine$alpineVer"
+		fi
+		export from alpineVer
+
+		case "$variant" in
+			fpm) cmd='["php-fpm"]' ;;
+			*) cmd='["php", "-a"]' ;;
+		esac
+		export cmd
+
+		echo "processing $version/$dir ..."
+		mkdir -p "$version/$dir"
+
+		{
+			generated_warning
+			gawk -f "$jqt" 'Dockerfile-linux.template'
```

Comment:
How _needed_ is the use of `gawk`? OSX doesn't have it by default but I can install it if needed. Just curious for portability's sake.

Reply:
BTW, it worked for me locally once I installed `gawk`. I'm fine if we need it overall

Reply:
Turns out, gawk is needed in order to read the correct formatting for the template file. If I remove that, the template result (Dockerfile) doesn't look nice whatsoever.

I would say it is needed in this case. Since it is working fine for you, it is a win-win for all. If anything, we can just add in the `README` letting developers know to install `gawk` if they are using Mac OS.

Reply:
ok yeah a note in the README would be good